### PR TITLE
Df generalized region

### PR DESCRIPTION
**Pull request by dflanderka**

### CODE REVIEW

File: src/fields/field.impl.hh [hunk 1]
```diff
@@ -328,12 +328,17 @@ void Field<spacedim,Value>::update_history(const TimeStep &time) {
 				}
 
 			} else if (shared_->list_it_->opt_val("region", domain_name)) {
+				domain = mesh()->region_db().get_region_set(domain_name);
```
Comment (jbrezmorf) [inline src/fields/field.impl.hh]: Should remove 'r_set' key.

File: src/mesh/elements.h [hunk 2]
```diff
@@ -126,6 +126,7 @@ class Element
     unsigned int dim_;
 
     friend class GmshMeshReader;
+    friend class Mesh;
```
Comment (jbrezmorf) [inline src/mesh/elements.h]: Why?

File: src/mesh/mesh.cc [hunk 3]
```diff
@@ -219,15 +221,18 @@ void Mesh::init_from_input() {
     Input::Array region_list;
     RegionDB::MapElementIDToRegionID el_to_reg_map;
 
-    // create regions from our input
-    if (in_record_.opt_val("regions", region_list)) {
-        region_db_.read_regions_from_input(region_list, el_to_reg_map);
-    }
     // read raw mesh, add regions from GMSH file
     GmshMeshReader reader( in_record_.val<FilePath>("mesh_file") );
     reader.read_mesh(this, &el_to_reg_map);
```
Comment (jbrezmorf) [inline src/mesh/mesh.cc]: So, now we can remove el_to_reg_map parameter and appropriate code form read_mesh.

File: src/mesh/region.cc [hunk 4]
```diff
@@ -146,6 +146,12 @@ Region RegionDB::add_region(unsigned int id, const std::string &label) {
         THROW(ExcNonuniqueLabel() << EI_Label(label) << EI_ID(id) << EI_IDOfOtherLabel(it_label->get_id()) );
     }
 
+    OnlyIDIter it_only_id = region_set_.get<OnlyID>().find(id);
```
Comment (jbrezmorf) [inline src/mesh/region.cc]: Why we need this if regions for region sets are added in close()?
However in order to simplify operations for creating region sets (union, etc.) we should
create region sets for regions just after reading the mesh.

Possibly we can just rename this method to rename_region, corresponding to the operation to name the region.

File: src/mesh/region.cc [hunk 5]
```diff
@@ -234,14 +240,19 @@ void RegionDB::close() {
...
    for(unsigned int i=0; i< size(); i++) {
        Region reg(i, *this);
+   	   RegionSet region_set;
+   	   region_set.push_back( reg );
+
 
        if (reg.is_boundary() && (reg.boundary_idx() < boundary_size()) ) {
            add_to_set("BOUNDARY", reg );
            add_to_set("ALL", reg);
+       	   add_set(reg.label(), region_set);
```
Comment (jbrezmorf) [inline src/mesh/region.cc]: Possibly we need to add regions into sets earlier.

File: src/mesh/region.cc [hunk 6]
```diff
@@ -535,6 +546,22 @@ Region RegionDB::replace_region_dim(DimIDIter it_undef_dim, unsigned int dim, bo
...
+	unsigned int index = it_only_id->index;
+
+	RegionItem item(index, it_only_id->get_id(), label, it_only_id->dim());
+	region_set_.replace(
+            region_set_.get<Index>().find(index),
+            item);
+
+	Region r_id=Region(index, *this);
+    // check boundary
+    if ( r_id.is_boundary() != boundary )
```
Comment (jbrezmorf) [inline src/mesh/region.cc]: Is it necessary to check the boundary at this level?

**Comment by jbrezmorf** [comment]
- This pull contains also changes in initialization of region DB which should be rather part of other pull for the mesh input changes.
- Missing remove of r_set key.
- Missing appropriate transformation file and changes in yaml files.

File: src/mesh/mesh.cc (hunk 3 above)
Comment (dflanderka) [inline src/mesh/mesh.cc]: We can't remove this code, it relates with key 'element_list', that is not removed or changed (maybe some changes will be in next issue #386). Marked 'if (in_record_.opt_val("regions" ...' is not removed, only moved after call setup_topology.

File: src/mesh/region.cc (hunk 4 above)
Comment (dflanderka) [inline src/mesh/region.cc]: region_set_ is RegionTable (database of all region) - variable name is misleading.
I propose to rename region_set_ (e.g. to region_table_).

Method is renamed.

File: src/mesh/region.cc (hunk 5 above)
Comment (dflanderka) [inline src/mesh/region.cc]: Problem with earlier added regions is with regions renamed by user. In this case we would must rename appropriate set too.

File: src/mesh/mesh.cc [hunk 7]
```diff
@@ -238,6 +243,20 @@ void Mesh::init_from_input() {
 
 
 
+void Mesh::modify_element_ids(const RegionDB::MapElementIDToRegionID *map) {
+	if (map) {
+		FOR_ELEMENTS(this, ele) {
```
Comment (jbrezmorf) [inline src/mesh/mesh.cc]: Try to use element.find_id(id) to loop only over the map.

**Pull request merged**